amethyst_animation/src/skinning/systems.rs

```diff
@@ -39,33 +34,34 @@ impl<'a> System<'a> for VertexSkinningSystem {
     type SystemData = (
         ReadStorage<'a, Joint>,
         ReadStorage<'a, GlobalTransform>,
-        ReadStorage<'a, Skin>,
+        WriteStorage<'a, Skin>,
         WriteStorage<'a, JointTransforms>,
     );
 
-    fn run(&mut self, (joints, transforms, skins, mut matrices): Self::SystemData) {
+    fn run(&mut self, (joints, transforms, mut skins, mut matrices): Self::SystemData) {
         self.updated.clear();
         transforms.populate_modified(&mut self.updated_id.as_mut().unwrap(), &mut self.updated);
         transforms.populate_inserted(&mut self.inserted_id.as_mut().unwrap(), &mut self.updated);
         self.updated_skins.clear();
         for (_, joint) in (&self.updated, &joints).join() {
-            self.updated_skins.add(joint.skin.id());
+            for skin in &joint.skins {
+                self.updated_skins.add(skin.id());
+            }
         }
 
-        for (_id, skin) in (&self.updated_skins, &skins).join() {
+        for (_id, skin) in (&self.updated_skins, &mut skins).join() {
             // Compute the joint transforms
-            self.joint_matrices.clear();
-            self.joint_matrices.extend(
+            skin.joint_matrices.clear();
+            let bind_shape = skin.bind_shape_matrix;
+            skin.joint_matrices.extend(
                 skin.joints
                     .iter()
-                    .map(|joint_entity| {
-                        (
-                            joints.get(*joint_entity).unwrap(),
-                            transforms.get(*joint_entity).unwrap(),
-                        )
+                    .zip(skin.inverse_bind_matrices.iter())
+                    .map(|(joint_entity, inverse_bind_matrix)| {
+                        (transforms.get(*joint_entity).unwrap(), inverse_bind_matrix)
```

Comment:
Optionnal: add a comment explaining how this unwrap is almost always safe.
Also to prevent denial of services where a server could crash a client, i think the almost always safe wont be enough and we will need proper handling.

Reply:
I think server security will require a full audit of the engine anyway, so while I think the comment is important, we don't have to handle all attack vectors yet.